Adding the MFT module with a basic QC task.

tomas-herman opened the pull request:
I am adding a basic QC task we prepared for the MFT. Everything I did should be contained in the Modules/MFT folder and adding this folder to the CMakeList.txt. I am also doing this pull request to learn how to contribute to the official repository.

knopers8 commented on Modules/MFT/include/MFT/BasicMFTDigitQcTask.h:
```diff
@@ -0,0 +1,57 @@
+  void reset() override;
+
+ private:
+  std::unique_ptr<TH1F> MFT_chip_index_H = nullptr;
```
For class member variables, please use the camel case convention with `m` at the beginning, as specified in the coding guidelines https://rawgit.com/AliceO2Group/CodingGuidelines/master/naming_formatting.html?showone=Variable_Names#Variable_Names

tomas-herman replied:
Done, MFT_chip_index_H replaced by mMFT_chip_index_H.

knopers8 commented on Modules/MFT/src/runDigitsRootFileReaderMFT.cxx:
```diff
@@ -0,0 +1,144 @@
+/// \endcode
+///
+
+// is this needed?
```
It isn't

knopers8 commented on Modules/MFT/src/runDigitsRootFileReaderMFT.cxx:
```diff
@@ -0,0 +1,144 @@
+
+
+// root headers
+#include "TFile.h"
```
Please use <> brackets to include anything that is outside of the QC project. With that, we reduce the compilation time a little bit by telling the compiler, that it should find these headers in external libraries.

tomas-herman replied:
Done.

knopers8 commented on Modules/MFT/src/runDigitsRootFileReaderMFT.cxx:
```diff
@@ -0,0 +1,144 @@
+    // fill in the message
+    //   LOG(INFO) << " runDigitsRootFileReaderMFT::run. In this ROF there are  " << DigitsInROF.size() << " digits";
+    pc.outputs().snapshot(Output{"MFT", "DIGITS", 0, Lifetime::Timeframe}, DigitsInROF);
+    pc.outputs().snapshot(Output{"MFT", "MFTDigitROF", 0, Lifetime::Timeframe}, oneROFvec);
```
Since these vectors are destructed anyway when going outside of the scope of this functions, consider using `adopt` instead of `snapshot`. The first one takes a pointer to your object, takes ownership and deletes it on its own (so you have to allocate it dynamically) - no unnecessary copies involved. The latter copies the object and sends that further.
You might want to try it out, if you care about performance here.

tomas-herman replied:
Thanks, I implemented the suggestion.

tomas-herman replied:
When I try to change snapshot for adopt I get an error when compiling QC. Here is the log: https://cernbox.cern.ch/index.php/s/rv2rKED9R6rqtjX

Would you know how to fix it?

knopers8 replied:
`adopt` needs a pointer, not a value. So first, you have to allocate your vector using `new`, so:
```
std::vector<o2::itsmft::Digit>* DigitsInROF = new std::vector<o2::itsmft::Digit>();
```
instead of
```
std::vector<o2::itsmft::Digit> DigitsInROF;
```
And then you have to pass that pointer to `adopt`.
If that doesn't compile, then it is a bug in `DataAllocator` and we can complain to its authors.

tomas-herman replied:
I have allocated the vector as you suggested and also I am using *DigitsInROF to refer to the vector. But I still seem to miss something as I get an error when using adopt. 

Here is the log file: https://cernbox.cern.ch/index.php/s/anJgUh6S09BMQM8
Here is the modified macro:https://cernbox.cern.ch/index.php/s/PX2Gax5m76ZHavu

knopers8 replied:
The code you have sent me has still `snapshot`, not `adopt`.
But by looking at the logs, the problem is that you are still passing a *value*, not a *pointer*. You do:
```
pc.outputs().adopt(Output{ "MFT", "DIGITS", 0, Lifetime::Timeframe }, *DigitsInROF);
```
As `DigitsInROF` is a pointer now. if you write `*DigitsInROF`, you get the value under that pointer.
Try removing the asterisk, then you will pass the pointer.

tomas-herman replied:
Sorry for the mess up with the file. I tried what you suggested but it also fails. 

Here is the, hopefully correct, file: https://cernbox.cern.ch/index.php/s/PX2Gax5m76ZHavu
Here is the new log: https://cernbox.cern.ch/index.php/s/kYb57xPDtoScqyv

knopers8 replied:
@matthiasrichter is `adopt` expected to work here? we are trying to do:
```
std::vector<o2::itsmft::ROFRecord>* oneROFvec = new std::vector<o2::itsmft::ROFRecord>();
std::vector<o2::itsmft::Digit>* DigitsInROF = new std::vector<o2::itsmft::Digit>();
...
pc.outputs().adopt(Output{ "MFT", "DIGITS", 0, Lifetime::Timeframe }, DigitsInROF);
pc.outputs().adopt(Output{ "MFT", "MFTDigitROF", 0, Lifetime::Timeframe }, oneROFvec);
```
but the compilation fails as in the logs from the comment above.

knopers8 replied:
Ok, since it is hard to get anyone's response recently, let's leave it as it is for now. We can improve it when the problem with `adopt` is understood.

tomas-herman replied:
Ok, thank you for all the help.

knopers8 commented on Modules/MFT/src/runDigitsRootFileReaderMFT.cxx:
```diff
@@ -0,0 +1,144 @@
+}; // end class definition
+
+
+// clang-format off
```
We use this if clang-format really breaks the formatting of complex structures. Please try to see if it breaks it in your case if you don't use `clang-format off`

tomas-herman replied:
Removed clang-format off.

tomas-herman commented on Modules/MFT/include/MFT/BasicMFTDigitQcCheck.h:
```diff
@@ -0,0 +1,46 @@
+
+///
+/// \file   BasicMFTDigitQcCheck.h
+/// \author Piotr Konopka
```
I've put my name and also Guillermo as we are both working on this.

tomas-herman commented on Modules/MFT/include/MFT/BasicMFTDigitQcCheck.h:
```diff
@@ -0,0 +1,46 @@
+/// \brief  Check whether a plot is empty or not.
+///
+/// \author Barthelemy von Haller
+class BasicMFTDigitQcCheck : public o2::quality_control::checker::CheckInterface
```
I've changed the name to be without MFT but I wanted to ask how do I delete the old object from the https://qcg-test.cern.ch/?page=objectTree , now that I will not need it any more?

knopers8 replied:
@Barthelemy What is the safest way to do that, so we don't delete too much by accident?

tomas-herman commented on Modules/MFT/include/MFT/BasicMFTDigitQcTask.h:
```diff
@@ -0,0 +1,57 @@
+#include "TH1F.h"
+#include "QualityControl/TaskInterface.h"
+
+class TH1F;
```
Done

tomas-herman commented on Modules/MFT/src/BasicMFTDigitQcTask.cxx:
```diff
@@ -0,0 +1,89 @@
+  MFT_chip_index_H = std::make_unique<TH1F>("MFT_chip_index", "MFT_chip_index", 936,-0.5,935.5);
+  getObjectsManager()->startPublishing(MFT_chip_index_H.get());
+  getObjectsManager()->addMetadata(MFT_chip_index_H->GetName(), "custom", "34");
+  getObjectsManager()->addCheck(MFT_chip_index_H.get(), "BasicMFTDigitQcCheck", "o2::quality_control_modules::mft::BasicMFTDigitQcCheck", "QcMFT");
```
Removed

tomas-herman commented:
Regarding the failing builds, the correct path to the file is: O2/Detectors/ITSMFT/common/base/include/ITSMFTBase/Digit.h
Similarly for another header: O2/DataFormats/Detectors/ITSMFT/common/include/DataFormatsITSMFT/ROFRecord.h
And several other headers in: O2/Framework/Core/include/Framework/xxx.h

How should I correctly write the path to the headers so that the compiler does not fail?

tomas-herman commented:
I am trying to install the clang-format tool following the guide you provided (https://github.com/AliceO2Group/CodingGuidelines#formatting-tool) so that I can implement the formatting rules but I get an error when trying to build (aliBuild build --defaults o2 Clang) 
Here is the log file: https://cernbox.cern.ch/index.php/s/mmMHuMIVm37cMrU

Could you help me? Or should I create a question in ALICE Talk?

knopers8 commented:
> How should I correctly write the path to the headers so that the compiler does not fail?

Please try using `target_link_libraries(QcMFT PUBLIC QualityControl O2::DataFormatsITSMFT)` instead of `target_link_libraries(QcMFT PUBLIC QualityControl)` in your CMakeLists.txt. It is probably that library that is missing. The O2::framework is probably not needed, since the QualityControl library brings it anyway.

> I am trying to install the clang-format tool following the guide you provided

Hmm, cool, it's an internal compiler error:
```
[ 36%] Building CXX object tools/clang/lib/Driver/CMakeFiles/obj.clangDriver.dir/Job.cpp.o
c++: internal compiler error: Killed (program cc1plus)
Please submit a full bug report,
with preprocessed source if appropriate.
See <file:///usr/share/doc/gcc-7/README.Bugs> for instructions.
tools/clang/lib/ASTMatchers/Dynamic/CMakeFiles/obj.clangDynamicASTMatchers.dir/build.make:101: recipe for target 'tools/clang/lib/ASTMatchers/Dynamic/CMakeFiles/obj.clangDynamicASTMatchers.dir/Registry.cpp.o' failed
make[2]: *** [tools/clang/lib/ASTMatchers/Dynamic/CMakeFiles/obj.clangDynamicASTMatchers.dir/Registry.cpp.o] Error 4
CMakeFiles/Makefile2:39614: recipe for target 'tools/clang/lib/ASTMatchers/Dynamic/CMakeFiles/obj.clangDynamicASTMatchers.dir/all' failed
make[1]: *** [tools/clang/lib/ASTMatchers/Dynamic/CMakeFiles/obj.clangDynamicASTMatchers.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
```
Please make sure that you have the latest master branch of the alidist repo. If it still persists, then please write on alice-talk.

tomas-herman commented:
> Please try using `target_link_libraries(QcMFT PUBLIC QualityControl O2::DataFormatsITSMFT)` instead of `target_link_libraries(QcMFT PUBLIC QualityControl)` in your CMakeLists.txt. It is probably that library that is missing. The O2::framework is probably not needed, since the QualityControl library brings it anyway.

Thank you. I've done it, hopefully it will solve the problem.

> Please make sure that you have the latest master branch of the alidist repo and try compiling again. If it still persists, then please write on alice-talk.

I've updated the master branch of alidist and tried several times. But I always get an internal compiler error. Here are just a few of the logs from my tries. Should I post to alice-talk then?

1) https://cernbox.cern.ch/index.php/s/eWYrvMUWfEvms4L
2) https://cernbox.cern.ch/index.php/s/uYjmJPAqyK8hBev
3) https://cernbox.cern.ch/index.php/s/Fy3ipl07ngPYHPM
4) https://cernbox.cern.ch/index.php/s/MtBGKDCV7uPtTn4
5) https://cernbox.cern.ch/index.php/s/dfF416pI5l2TAew

Also since I updated alidist I had to update O2 because I was not able to compile O2/QC any more. While compiling O2 I got the internal compiler error many times. But I always just run the build again, sometimes I also put alibuild clean in between the builds and by some black magic after few hours of trying this, it worked. Would you know why?

knopers8 commented:
> I've updated the master branch of alidist and tried several times. But I always get an internal compiler error. Here are just a few of the logs from my tries. Should I post to alice-talk then?

Yes please. I have no idea what is going on there. I am sorry about that.

> Also since I updated alidist I had to update O2 because I was not able to compile O2/QC any more. While compiling O2 I got the internal compiler error many times. But I always just run the build again, sometimes I also put alibuild clean in between the builds and by some black magic after few hours of trying this, it worked. Would you know why?

Not really. It might have something to do with cached bulids, which are not rebuilt when necessary. When the whole build starts to go out of control and I don't have time to investigate it, I usually delete the `sw` directory and let it rebuild everything from scratch.

tomas-herman commented:
After getting help on alice-talk, I was able to install clang using the guide and I used the command `find . -iname "*.h" -o -iname "*.cpp" | xargs clang-format -style=file -i ` in the MFT module folder. But from the checks it seems it was not enough? What else has to be done to implement the formatting correctly?

knopers8 commented:
Now it is just CI that failed (you can click the Details link to see it):
```
resolvconf

Adding APT Sources

Disallowing sources: llvm-toolchain-xenial-7

To add unlisted APT sources, follow instructions in https://docs.travis-ci.com/user/installing-dependencies#Installing-Packages-with-the-APT-Addon

11.05s$ travis_apt_get_update

Installing APT Packages

0.30s$ sudo -E apt-get -yq --no-install-suggests --no-install-recommends $(travis_apt_get_options) install clang-format-7

Reading package lists...

Building dependency tree...

Reading state information...

E: Unable to locate package clang-format-7
apt-get.diagnostics

apt-get install failed

The command "sudo -E apt-get -yq --no-install-suggests --no-install-recommends $(travis_apt_get_options) install clang-format-7" failed and exited with 100 during .

Your build has been stopped.
```
@Barthelemy Could you have a look?

tomas-herman commented:
Did the build get stuck?

knopers8 commented:
Yes, I will trigger it again.

tomas-herman commented:
Isn't it taking too long again? It is strange, because when I pushed the last commit all the builds were successful.

knopers8 commented:
@MichaelLettrich The o2 build is stuck, could you have a look?

MichaelLettrich commented:
Hi we had problems with the IT VM infrastructure. I will have a look and restart things.

knopers8 commented:
Thanks

tomas-herman commented:
Thank you.
